[UI-Side compositing] Move ScrollerPairMac to ScrollingTreeScrollingNodeDelegateMac

nmoucht opened the pull request:
#### 20982a6f9b4d0fe71fefa63225ce1383995a3126
<pre>
[UI-Side compositing] Move ScrollerPairMac to ScrollingTreeScrollingNodeDelegateMac
<a href="https://bugs.webkit.org/show_bug.cgi?id=251297">https://bugs.webkit.org/show_bug.cgi?id=251297</a>
rdar://104762822

Reviewed by Simon Fraser.

Move ScrollerPairMac and ScrollerMac to WebCore so they can be used in ScrollingTreeScrollingNodeDelegateMac. Move the
scrollbar code in ScrollingTreeFrameScrollingNodeRemoteMac and ScrollingTreeOverflowScrollingNodeRemoteMac to
ScrollingTreeScrollingNodeDelegateMac. Remove m_verticalScrollerImp and m_horizontalScrollerImp so that
the non UI-side compositing code uses ScrollerPairMac as well.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/scrolling/mac/ScrollerMac.h: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerMac.h.
(WebCore::ScrollerMac::pair):
(WebCore::ScrollerMac::orientation const):
(WebCore::ScrollerMac::hostLayer const):
(WebCore::ScrollerMac::scrollerImp):
(WebCore::ScrollerMac::setscrollerImp):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerMac.mm.
(-[WebScrollbarPartAnimationMac initWithScroller:featureToAnimate:animateFrom:animateTo:duration:]):
(-[WebScrollbarPartAnimationMac startAnimation]):
(-[WebScrollbarPartAnimationMac setStartValue:]):
(-[WebScrollbarPartAnimationMac setEndValue:]):
(-[WebScrollbarPartAnimationMac setCurrentProgress:]):
(-[WebScrollbarPartAnimationMac invalidate]):
(-[WebScrollerImpDelegateMac initWithScroller:]):
(-[WebScrollerImpDelegateMac cancelAnimations]):
(-[WebScrollerImpDelegateMac convertRectToBacking:]):
(-[WebScrollerImpDelegateMac convertRectFromBacking:]):
(-[WebScrollerImpDelegateMac layer]):
(-[WebScrollerImpDelegateMac mouseLocationInScrollerForScrollerImp:]):
(-[WebScrollerImpDelegateMac convertRectToLayer:]):
(-[WebScrollerImpDelegateMac shouldUseLayerPerPartForScrollerImp:]):
(-[WebScrollerImpDelegateMac effectiveAppearanceForScrollerImp:]):
(-[WebScrollerImpDelegateMac setUpAlphaAnimation:featureToAnimate:animateAlphaTo:duration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateKnobAlphaTo:duration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateTrackAlphaTo:duration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateUIStateTransitionWithDuration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateExpansionTransitionWithDuration:]):
(-[WebScrollerImpDelegateMac scrollerImp:overlayScrollerStateChangedTo:]):
(-[WebScrollerImpDelegateMac invalidate]):
(WebCore::ScrollerMac::ScrollerMac):
(WebCore::ScrollerMac::~ScrollerMac):
(WebCore::ScrollerMac::attach):
(WebCore::ScrollerMac::setHostLayer):
(WebCore::ScrollerMac::updateValues):
(WebCore::ScrollerMac::convertFromContent const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerPairMac.h.
(WebCore::ScrollerPairMac::verticalScroller):
(WebCore::ScrollerPairMac::horizontalScroller):
(WebCore::ScrollerPairMac::lastKnownMousePosition const):
(WebCore::ScrollerPairMac::scrollerImpPair):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerPairMac.mm.
(-[WebScrollerImpPairDelegateMac initWithScrollerPair:]):
(-[WebScrollerImpPairDelegateMac invalidate]):
(-[WebScrollerImpPairDelegateMac contentAreaRectForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac inLiveResizeForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac mouseLocationInContentAreaForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:setContentAreaNeedsDisplayInRect:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:updateScrollerStyleForNewRecommendedScrollerStyle:]):
(WebCore::ScrollerPairMac::ScrollerPairMac):
(WebCore::ScrollerPairMac::init):
(WebCore::ScrollerPairMac::~ScrollerPairMac):
(WebCore::ScrollerPairMac::handleWheelEvent):
(WebCore::ScrollerPairMac::handleMouseEvent):
(WebCore::ScrollerPairMac::updateValues):
(WebCore::ScrollerPairMac::visibleSize const):
(WebCore::ScrollerPairMac::useDarkAppearance const):
(WebCore::ScrollerPairMac::valuesForOrientation):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::releaseReferencesToScrollerImpsOnTheMainThread):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::initScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarLayers):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEventForScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleMouseEventForScrollbars):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::ScrollingTreeFrameScrollingNodeRemoteMac):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::commitStateBeforeChildren):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::repositionRelatedLayers):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::handleWheelEvent):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::handleMouseEvent):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeOverflowScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::ScrollingTreeOverflowScrollingNodeRemoteMac):
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::commitStateBeforeChildren):
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::repositionRelatedLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeOverflowScrollingNodeRemoteMac.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/259592@main">https://commits.webkit.org/259592@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b23c61f42422b152aa0c705791290d59d94b36ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105329 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14405 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38210 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114586 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/174770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109232 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15585 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5334 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97638 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111087 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95034 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93919 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26668 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7730 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28028 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7839 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13874 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47572 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9626 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3535 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
```diff
@@ -76,8 +82,7 @@ class ScrollingTreeScrollingNodeDelegateMac final : public ThreadedScrollingTree
 
-    RetainPtr<NSScrollerImp> m_verticalScrollerImp;
-    RetainPtr<NSScrollerImp> m_horizontalScrollerImp;
+    std::unique_ptr<ScrollerPairMac> m_scrollerPair;
```
Does it have to be a std::unique_ptr or can we just store by value?

nmoucht replied:
Changed to store by value.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -76,8 +94,8 @@
     if (wasInMomentumPhase != m_inMomentumPhase) {
-        [m_verticalScrollerImp setUsePresentationValue:m_inMomentumPhase];
-        [m_horizontalScrollerImp setUsePresentationValue:m_inMomentumPhase];
+        [m_scrollerPair->verticalScroller().scrollerImp() setUsePresentationValue:m_inMomentumPhase];
```
Maybe a function on `ScrollerPairMac` can wrap these two calls.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -267,24 +285,24 @@
 void ScrollingTreeScrollingNodeDelegateMac::updateScrollbarPainters()
 {
-    if (m_inMomentumPhase && (m_verticalScrollerImp || m_horizontalScrollerImp)) {
+    if (m_inMomentumPhase && (m_scrollerPair->verticalScroller().scrollerImp() || m_scrollerPair->horizontalScroller().scrollerImp())) {
```
Maybe add a helper on `ScrollerPairMac` to ask if there's at least one scrollerImp

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -294,14 +312,38 @@
         // FIXME: This is a workaround in place for the time being since NSScrollerImps cannot be deallocated
         // on a non-main thread. rdar://problem/24535055
-        WTF::callOnMainThread([verticalScrollerImp = WTFMove(m_verticalScrollerImp), horizontalScrollerImp = WTFMove(m_horizontalScrollerImp)] {
+        WTF::callOnMainThread([verticalScrollerImp = m_scrollerPair->verticalScroller().scrollerImp(), horizontalScrollerImp = m_scrollerPair->horizontalScroller().scrollerImp()] {
```
This needs to "take" the scrollerImps (they need to be nulled out on the pair), so that the last ref goes away on the main thread. Maybe ScrollerPairMac itself can take care of this bounce to the main thread.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -294,14 +312,38 @@
 
+void ScrollingTreeScrollingNodeDelegateMac::initScrollbars()
+{
+    m_scrollerPair->init();
```
Not sure why we need the `init` separate from construction?

nmoucht replied:
This is to avoid unnecessarily allocating the member variables that only ui-side compositing uses with ui-side compositing off.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -294,14 +312,38 @@
+
+void ScrollingTreeScrollingNodeDelegateMac::updateScrollbarLayers()
+{
+    if (m_scrollerPair)
```
Why the null test here but not elsewhere? Aren't we guaranteed that the pair is non-null?

nmoucht replied:
Removed these checks.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.h:
```diff
@@ -55,7 +55,10 @@ class ScrollerMac {
     CALayer *hostLayer() const { return m_hostLayer.get(); }
     void setHostLayer(CALayer *);
 
+    RetainPtr<NSScrollerImp> takeScrollerImp() { return WTFMove(m_scrollerImp); }
```
Better to use `std::exchange(m_scrollerImp, { })`;

nmoucht replied:
Fixed.

nmoucht committed:
[UI-Side compositing] Move ScrollerPairMac to ScrollingTreeScrollingNodeDelegateMac
https://bugs.webkit.org/show_bug.cgi?id=251297
rdar://104762822

Reviewed by Simon Fraser.

Move ScrollerPairMac and ScrollerMac to WebCore so they can be used in ScrollingTreeScrollingNodeDelegateMac. Move the
scrollbar code in ScrollingTreeFrameScrollingNodeRemoteMac and ScrollingTreeOverflowScrollingNodeRemoteMac to
ScrollingTreeScrollingNodeDelegateMac. Remove m_verticalScrollerImp and m_horizontalScrollerImp so that
the non UI-side compositing code uses ScrollerPairMac as well.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/scrolling/mac/ScrollerMac.h: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerMac.h.
(WebCore::ScrollerMac::pair):
(WebCore::ScrollerMac::orientation const):
(WebCore::ScrollerMac::hostLayer const):
(WebCore::ScrollerMac::scrollerImp):
(WebCore::ScrollerMac::setscrollerImp):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerMac.mm.
(-[WebScrollbarPartAnimationMac initWithScroller:featureToAnimate:animateFrom:animateTo:duration:]):
(-[WebScrollbarPartAnimationMac startAnimation]):
(-[WebScrollbarPartAnimationMac setStartValue:]):
(-[WebScrollbarPartAnimationMac setEndValue:]):
(-[WebScrollbarPartAnimationMac setCurrentProgress:]):
(-[WebScrollbarPartAnimationMac invalidate]):
(-[WebScrollerImpDelegateMac initWithScroller:]):
(-[WebScrollerImpDelegateMac cancelAnimations]):
(-[WebScrollerImpDelegateMac convertRectToBacking:]):
(-[WebScrollerImpDelegateMac convertRectFromBacking:]):
(-[WebScrollerImpDelegateMac layer]):
(-[WebScrollerImpDelegateMac mouseLocationInScrollerForScrollerImp:]):
(-[WebScrollerImpDelegateMac convertRectToLayer:]):
(-[WebScrollerImpDelegateMac shouldUseLayerPerPartForScrollerImp:]):
(-[WebScrollerImpDelegateMac effectiveAppearanceForScrollerImp:]):
(-[WebScrollerImpDelegateMac setUpAlphaAnimation:featureToAnimate:animateAlphaTo:duration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateKnobAlphaTo:duration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateTrackAlphaTo:duration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateUIStateTransitionWithDuration:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateExpansionTransitionWithDuration:]):
(-[WebScrollerImpDelegateMac scrollerImp:overlayScrollerStateChangedTo:]):
(-[WebScrollerImpDelegateMac invalidate]):
(WebCore::ScrollerMac::ScrollerMac):
(WebCore::ScrollerMac::~ScrollerMac):
(WebCore::ScrollerMac::attach):
(WebCore::ScrollerMac::setHostLayer):
(WebCore::ScrollerMac::updateValues):
(WebCore::ScrollerMac::convertFromContent const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerPairMac.h.
(WebCore::ScrollerPairMac::verticalScroller):
(WebCore::ScrollerPairMac::horizontalScroller):
(WebCore::ScrollerPairMac::lastKnownMousePosition const):
(WebCore::ScrollerPairMac::scrollerImpPair):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollerPairMac.mm.
(-[WebScrollerImpPairDelegateMac initWithScrollerPair:]):
(-[WebScrollerImpPairDelegateMac invalidate]):
(-[WebScrollerImpPairDelegateMac contentAreaRectForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac inLiveResizeForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac mouseLocationInContentAreaForScrollerImpPair:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:setContentAreaNeedsDisplayInRect:]):
(-[WebScrollerImpPairDelegateMac scrollerImpPair:updateScrollerStyleForNewRecommendedScrollerStyle:]):
(WebCore::ScrollerPairMac::ScrollerPairMac):
(WebCore::ScrollerPairMac::init):
(WebCore::ScrollerPairMac::~ScrollerPairMac):
(WebCore::ScrollerPairMac::handleWheelEvent):
(WebCore::ScrollerPairMac::handleMouseEvent):
(WebCore::ScrollerPairMac::updateValues):
(WebCore::ScrollerPairMac::visibleSize const):
(WebCore::ScrollerPairMac::useDarkAppearance const):
(WebCore::ScrollerPairMac::valuesForOrientation):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::releaseReferencesToScrollerImpsOnTheMainThread):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::initScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarLayers):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEventForScrollbars):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleMouseEventForScrollbars):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::ScrollingTreeFrameScrollingNodeRemoteMac):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::commitStateBeforeChildren):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::repositionRelatedLayers):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::handleWheelEvent):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::handleMouseEvent):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeOverflowScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::ScrollingTreeOverflowScrollingNodeRemoteMac):
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::commitStateBeforeChildren):
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::repositionRelatedLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeOverflowScrollingNodeRemoteMac.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/259592@main
